File: frontend/src/components/canvas/nodeRegistry.ts
export interface NodeMeta {
  op_type: string;
  label: string;
  category: string;
  defaultParams: Record<string, any>;
  inputs?: number;
  outputs?: number;
  color?: string;
  description?: string;
}

export const NODE_CATEGORIES = [
  { id: "io", label: "输入输出" },
  { id: "conv", label: "卷积层" },
  { id: "norm", label: "归一化" },
  { id: "activation", label: "激活函数" },
  { id: "pool", label: "池化层" },
  { id: "linear", label: "全连接层" },
  { id: "structure", label: "结构层" },
  { id: "multi", label: "多输入" },
];

const NODE_REGISTRY: Record<string, NodeMeta> = {
  "Input": {
    op_type: "Input",
    label: "Input",
    category: "io",
    defaultParams: { shape: [3, 224, 224] },
    outputs: 1,
    color: "#6366f1",
    description: "输入层，定义输入张量的形状 [C, H, W]",
  },
  "Output": {
    op_type: "Output",
    label: "Output",
    category: "io",
    defaultParams: {},
    inputs: 1,
    color: "#6366f1",
    description: "输出层，标记模型的最终输出",
  },
  "Conv2d": {
    op_type: "Conv2d",
    label: "Conv2d",
    category: "conv",
    defaultParams: {
      in_channels: 3,
      out_channels: 64,
      kernel_size: 3,
      stride: 1,
      padding: 1,
      dilation: 1,
      groups: 1,
      bias: true,
    },
    inputs: 1,
    outputs: 1,
    color: "#10b981",
    description: "2D 卷积层，用于提取图像特征",
  },
  "DepthwiseConv2d": {
    op_type: "DepthwiseConv2d",
    label: "DepthwiseConv2d",
    category: "conv",
    defaultParams: {
      in_channels: 64,
      kernel_size: 3,
      stride: 1,
      padding: 1,
    },
    inputs: 1,
    outputs: 1,
    color: "#10b981",
    description: "深度可分离卷积，参数量更少",
  },
  "ConvTranspose2d": {
    op_type: "ConvTranspose2d",
    label: "ConvTranspose2d",
    category: "conv",
    defaultParams: {
      in_channels: 64,
      out_channels: 3,
      kernel_size: 3,
      stride: 1,
      padding: 1,
      output_padding: 0,
    },
    inputs: 1,
    outputs: 1,
    color: "#10b981",
    description: "转置卷积，用于上采样和生成任务",
  },
  "BatchNorm2d": {
    op_type: "BatchNorm2d",
    label: "BatchNorm2d",
    category: "norm",
    defaultParams: {
      num_features: 64,
      eps: 1e-5,
      momentum: 0.1,
      affine: true,
    },
    inputs: 1,
    outputs: 1,
    color: "#f59e0b",
    description: "批归一化层，加速训练并提高稳定性",
  },
  "LayerNorm": {
    op_type: "LayerNorm",
    label: "LayerNorm",
    category: "norm",
    defaultParams: {
      normalized_shape: [64],
      eps: 1e-5,
    },
    inputs: 1,
    outputs: 1,
    color: "#f59e0b",
    description: "层归一化，常用于 Transformer",
  },
  "ReLU": {
    op_type: "ReLU",
    label: "ReLU",
    category: "activation",
    defaultParams: { inplace: true },
    inputs: 1,
    outputs: 1,
    color: "#ec4899",
    description: "ReLU 激活函数，引入非线性",
  },
  "LeakyReLU": {
    op_type: "LeakyReLU",
    label: "LeakyReLU",
    category: "activation",
    defaultParams: { negative_slope: 0.01, inplace: true },
    inputs: 1,
    outputs: 1,
    color: "#ec4899",
    description: "LeakyReLU，解决 ReLU 的神经元死亡问题",
  },
  "GELU": {
    op_type: "GELU",
    label: "GELU",
    category: "activation",
    defaultParams: {},
    inputs: 1,
    outputs: 1,
    color: "#ec4899",
    description: "GELU 激活函数，用于 Transformer",
  },
  "Sigmoid": {
    op_type: "Sigmoid",
    label: "Sigmoid",
    category: "activation",
    defaultParams: {},
    inputs: 1,
    outputs: 1,
    color: "#ec4899",
    description: "Sigmoid 激活函数，输出范围 (0, 1)",
  },
  "Tanh": {
    op_type: "Tanh",
    label: "Tanh",
    category: "activation",
    defaultParams: {},
    inputs: 1,
    outputs: 1,
    color: "#ec4899",
    description: "Tanh 激活函数，输出范围 (-1, 1)",
  },
  "MaxPool2d": {
    op_type: "MaxPool2d",
    label: "MaxPool2d",
    category: "pool",
    defaultParams: {
      kernel_size: 2,
      stride: 2,
      padding: 0,
    },
    inputs: 1,
    outputs: 1,
    color: "#8b5cf6",
    description: "最大池化，降采样并保留最强特征",
  },
  "AvgPool2d": {
    op_type: "AvgPool2d",
    label: "AvgPool2d",
    category: "pool",
    defaultParams: {
      kernel_size: 2,
      stride: 2,
      padding: 0,
    },
    inputs: 1,
    outputs: 1,
    color: "#8b5cf6",
    description: "平均池化，平滑特征",
  },
  "AdaptiveAvgPool2d": {
    op_type: "AdaptiveAvgPool2d",
    label: "AdaptiveAvgPool2d",
    category: "pool",
    defaultParams: {
      output_size: [1, 1],
    },
    inputs: 1,
    outputs: 1,
    color: "#8b5cf6",
    description: "自适应平均池化，输出固定尺寸",
  },
  "Linear": {
    op_type: "Linear",
    label: "Linear",
    category: "linear",
    defaultParams: {
      in_features: 512,
      out_features: 10,
      bias: true,
    },
    inputs: 1,
    outputs: 1,
    color: "#06b6d4",
    description: "全连接层，用于特征映射",
  },
  "Flatten": {
    op_type: "Flatten",
    label: "Flatten",
    category: "structure",
    defaultParams: {
      start_dim: 1,
      end_dim: -1,
    },
    inputs: 1,
    outputs: 1,
    color: "#64748b",
    description: "展平层，将多维张量展平为一维",
  },
  "Dropout": {
    op_type: "Dropout",
    label: "Dropout",
    category: "structure",
    defaultParams: {
      p: 0.5,
      inplace: false,
    },
    inputs: 1,
    outputs: 1,
    color: "#64748b",
    description: "Dropout，防止过拟合",
  },
  "Dropout2d": {
    op_type: "Dropout2d",
    label: "Dropout2d",
    category: "structure",
    defaultParams: {
      p: 0.5,
      inplace: false,
    },
    inputs: 1,
    outputs: 1,
    color: "#64748b",
    description: "2D Dropout，用于卷积网络",
  },
  "Add": {
    op_type: "Add",
    label: "Add",
    category: "multi",
    defaultParams: {},
    inputs: 2,
    outputs: 1,
    color: "#ef4444",
    description: "逐元素相加，用于残差连接",
  },
  "Concat": {
    op_type: "Concat",
    label: "Concat",
    category: "multi",
    defaultParams: {
      dim: 1,
    },
    inputs: 2,
    outputs: 1,
    color: "#ef4444",
    description: "拼接张量，沿指定维度连接",
  },
};

export function getNodeMeta(op_type: string): NodeMeta {
  return NODE_REGISTRY[op_type];
}

export function getAllNodes(): NodeMeta[] {
  return Object.values(NODE_REGISTRY);
}

File: backend/core/ir/codegen/node_registry.py
"""
算子注册表：维护每种 OpType 对应的 PyTorch 代码片段生成逻辑。

每个算子注册一个 builder 函数：
    (node: IRNode) -> (init_code: str, forward_code: str)

- init_code:    在 __init__ 中的初始化语句，例如 "nn.Conv2d(3, 64, 3)"
- forward_code: 在 forward 中的调用语句，例如 "x = self.conv1(x)"
"""

from typing import Callable
from core.ir.model_ir import IRNode, OpType

# 类型别名
NodeBuilder = Callable[[IRNode], tuple[str, str]]

_REGISTRY: dict[OpType, NodeBuilder] = {}

def register(op_type: OpType):
    """装饰器：注册某 OpType 对应的代码生成函数"""
    def decorator(fn: NodeBuilder) -> NodeBuilder:
        _REGISTRY[op_type] = fn
        return fn
    return decorator

def get_builder(op_type: OpType) -> NodeBuilder:
    if op_type not in _REGISTRY:
        raise NotImplementedError(f"OpType '{op_type}' 尚未注册代码生成器")
    return _REGISTRY[op_type]

# ─────────────────────────────────────────────
# 各算子的代码生成注册
# ─────────────────────────────────────────────

@register(OpType.INPUT)
def _input_builder(node: IRNode) -> tuple[str, str]:
    # Input 节点不生成 nn.Module，只是 forward 的入口
    return "", ""   # init 无需生成，forward 会单独处理

@register(OpType.OUTPUT)
def _output_builder(node: IRNode) -> tuple[str, str]:
    return "", ""

@register(OpType.CONV2D)
def _conv2d_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    in_ch = p.get("in_channels", 3)
    out_ch = p.get("out_channels", 64)
    ks = p.get("kernel_size", 3)
    stride = p.get("stride", 1)
    padding = p.get("padding", 1)
    groups = p.get("groups", 1)
    bias = p.get("bias", True)

    init = (
        f"nn.Conv2d({in_ch}, {out_ch}, kernel_size={ks}, "
        f"stride={stride}, padding={padding}, groups={groups}, bias={bias})"
    )
    return init, ""

@register(OpType.DEPTHWISE_CONV2D)
def _dw_conv2d_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    channels = p.get("in_channels", 64)
    ks = p.get("kernel_size", 3)
    stride = p.get("stride", 1)
    padding = p.get("padding", 1)

    init = (
        f"nn.Conv2d({channels}, {channels}, kernel_size={ks}, "
        f"stride={stride}, padding={padding}, groups={channels}, bias=False)"
    )
    return init, ""

@register(OpType.BATCH_NORM2D)
def _bn2d_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    num_features = p.get("num_features", 64)
    eps = p.get("eps", 1e-5)
    momentum = p.get("momentum", 0.1)
    init = f"nn.BatchNorm2d({num_features}, eps={eps}, momentum={momentum})"
    return init, ""

@register(OpType.LAYER_NORM)
def _ln_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    normalized_shape = p.get("normalized_shape", [64])
    init = f"nn.LayerNorm({normalized_shape})"
    return init, ""

@register(OpType.RELU)
def _relu_builder(node: IRNode) -> tuple[str, str]:
    return "nn.ReLU(inplace=True)", ""

@register(OpType.LEAKY_RELU)
def _leaky_relu_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    slope = p.get("negative_slope", 0.01)
    return f"nn.LeakyReLU(negative_slope={slope}, inplace=True)", ""

@register(OpType.GELU)
def _gelu_builder(node: IRNode) -> tuple[str, str]:
    return "nn.GELU()", ""

@register(OpType.SIGMOID)
def _sigmoid_builder(node: IRNode) -> tuple[str, str]:
    return "nn.Sigmoid()", ""

@register(OpType.TANH)
def _tanh_builder(node: IRNode) -> tuple[str, str]:
    return "nn.Tanh()", ""

@register(OpType.MAX_POOL2D)
def _maxpool_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    ks = p.get("kernel_size", 2)
    stride = p.get("stride", None)
    padding = p.get("padding", 0)
    stride_str = f", stride={stride}" if stride is not None else ""
    init = f"nn.MaxPool2d(kernel_size={ks}{stride_str}, padding={padding})"
    return init, ""

@register(OpType.AVG_POOL2D)
def _avgpool_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    ks = p.get("kernel_size", 2)
    stride = p.get("stride", None)
    stride_str = f", stride={stride}" if stride is not None else ""
    init = f"nn.AvgPool2d(kernel_size={ks}{stride_str})"
    return init, ""

@register(OpType.ADAPTIVE_AVG_POOL2D)
def _adaptive_avgpool_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    output_size = p.get("output_size", (1, 1))
    init = f"nn.AdaptiveAvgPool2d(output_size={output_size})"
    return init, ""

@register(OpType.LINEAR)
def _linear_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    in_f = p.get("in_features", 512)
    out_f = p.get("out_features", 10)
    bias = p.get("bias", True)
    init = f"nn.Linear({in_f}, {out_f}, bias={bias})"
    return init, ""

@register(OpType.FLATTEN)
def _flatten_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    start_dim = p.get("start_dim", 1)
    init = f"nn.Flatten(start_dim={start_dim})"
    return init, ""

@register(OpType.DROPOUT)
def _dropout_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    prob = p.get("p", 0.5)
    init = f"nn.Dropout(p={prob})"
    return init, ""

@register(OpType.DROPOUT2D)
def _dropout2d_builder(node: IRNode) -> tuple[str, str]:
    p = node.params
    prob = p.get("p", 0.5)
    init = f"nn.Dropout2d(p={prob})"
    return init, ""

@register(OpType.ADD)
def _add_builder(node: IRNode) -> tuple[str, str]:
    # Add 不需要 nn.Module，在 forward 里直接做张量加法
    return "", ""

@register(OpType.CONCAT)
def _concat_builder(node: IRNode) -> tuple[str, str]:
    # Concat 也不需要 nn.Module
    return "", ""